### ISSUE TITLE
Use AssemblyLoadContext rather than the AppDomain shim

### ISSUE THREAD

**Issue by seesharper** [filipw/dotnet-script#262]

**Comment by Diaskhan** [comment]
What kind positive behaviours provide 
AssemblyLoadContext? 

If there loading assembly of scripting? 
What if app gonna be long running processes ? It possible to unload assembly that scripting library compiled ? 

I know the unloading of assembly in work ? Any one has deal with that behaviours of net core, that assemblies is not unloadable?

It possible to replace script in memory with new version script ?

**Comment by filipw** [comment]
at the moment you can't unload, as .NET Core doesn't support it yet. However, it will be supported in future versions of .NET Core. see https://github.com/dotnet/coreclr/pull/18476

**Comment by Diaskhan** [comment]
Yeah i know! It seems to me if we use scripting memory stack will grow to the size of compiled sxript ?

**Comment by hrumhurum** [comment]
Can I work on this task?

The plan:

1. Change `AppDomain` usages to functionality provided by `AssemblyLoadContext`
2. Allow passing an optional `AssemblyLoadContext` instance in higher-level options such as `ExecuteScriptCommandOptions`. In this way, the assembly loading behavior can be customized

**Comment by seesharper** [comment]
This is something I did look into a couple of years ago and back then I could not find i way to pass an `AssemblyLoadContext` to the underlying roslyn scripting API's. Nor did I find a way to hack it through reflection either which we have done a couple of times :) Things might have changed since then though so maybe there is a way now 👍. 100% isolation from the host would be a great thing to have as it would solve version conflicts of shared libraries between the host and the scripting "environment".

**Comment by hrumhurum** [comment]
I was able to come up with a successful prototype implementation. Now working on a formal pull request.

**Comment by seesharper** [comment]
Really? Wow, that's impressive.👌
